### PR TITLE
[JENKINS-26535] added support for parameterized parameter types

### PR DESCRIPTION
**Pull request by daspilker**
Follow up of https://groups.google.com/forum/#!msg/jenkinsci-dev/-XVFCCiE4Bs

This fixed the issue mentioned on the mailing list, but does probably not work for all cases and might not fix JENKINS-26535 completely.

### CODE REVIEW

File: plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java]: Most of this is actually reindentation, making the diff look a lot bigger than it actually is. Had to apply the WS ignore setting in GitHub to see it.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java]: This seems to have been removed and I am not sure why. Where is the fallback?

Comment (jglick) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java]: I guess it is just not necessary any more and whatever sanity-checking was needed is already done in `DescribableModel`?

Comment (daspilker) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java]: The fallback became unreachable, so I removed it. I was surprised that no unit tests failed. There is not much sanity-checking in `DescribableModel` or `DescribableParameter`.

I thought about limiting support for `Homo`/`HeterogeneousObjectType` to `Describable` types and keeping the fallback, but that triggered a test failure in `DescribableModelTest#recursion` since `Recursion` is not a `Describable`. I'm not sure if that is on purpose.

**Comment by daspilker** [comment]
I pushed some tests for the problems that have been mentioned in the description of JENKINS-26535. So I'm pretty confident that this fixes JENKINS-26535.

I also ran the Job DSL tests successfully with a snapshot build.

**Review by jetersen — DISMISSED** [review]

**Comment by baymac** [comment]
@daspilker @jglick Is it ready for release?

File: plugin/src/test/java/org/jenkinsci/plugins/structs/describable/DescribableModelTest.java
Comment (casz) [inline plugin/src/test/java/org/jenkinsci/plugins/structs/describable/DescribableModelTest.java]: Could add
```suggestion
    @Test
    @Issue("JENKINS-26535")
```

**Comment by casz** [comment]
@abayer @dwnusbaum @jglick 👋 issue confirms that 26535 👍 

Would be great with a release!

**Comment by casz** [comment]
@car-roll this is still a blocker, can this be added to fix this very old bug

**Comment by casz** [comment]
ping @car-roll @jglick 😓

**Comment by jglick** [comment]
I do not maintain this plugin, do not ask me.

**Comment by casz** [comment]
You reviewed the code so maybe the maintainer is waiting for your approval? 😅

**Comment by jglick** [comment]
> parameters of type `Map` were not supported in some cases

Because `Map` is not a legal databinding type. The only compound types supported are arrays or `List`s whose element type is assignable to `Describable`.

**Comment by jglick** [comment]
The basic description is in https://github.com/jenkinsci/structs-plugin/blob/696fe7b7763fba23398ee5842db917597baf2014/plugin/src/main/java/org/jenkinsci/plugins/structs/describable/DescribableModel.java#L278-L281 though this likely needs to be expanded and clarified some.

**Comment by dwnusbaum** [comment]
> Because Map is not a legal databinding type. The only compound types supported are arrays or Lists whose element type is assignable to Describable.

@jglick Ok, well with this PR in its current state, they are supported. I guess this is related to the old fallback that throws an `UnsupportedOperationException` no longer being reachable as you noted in [this comment](https://github.com/jenkinsci/structs-plugin/pull/52#discussion_r315373192).

**Comment by casz** [comment]
👋 This is still blocking would be great to have a release 🙏

**Comment by oleg-nenashev** [comment]
Apparently I am listed as a plugin maintainer. I would be interested to cut the release if @jglick and @dwnusbaum are satisfied by the current state

**Comment by jglick** [comment]
I would leave it to @dwnusbaum.

**Comment by darxriggs** [comment]
It looks like this pull request is stuck. I therefore like to ask who could follow up on it?

**Comment by timja** [comment]
Hi @dwnusbaum, just taking a bit of a look at this,

The test case you suggested doesn't appear to be passing anymore on master,

```
java.lang.ArrayIndexOutOfBoundsException: 1
	at org.jenkinsci.plugins.structs.describable.DescribableModel.<init>(DescribableModel.java:160)
	at org.jenkinsci.plugins.structs.describable.DescribableModelTest.schema(DescribableModelTest.java:925)
	at org.jenkinsci.plugins.structs.describable.DescribableModelTest.mapParameter(DescribableModelTest.java:81)

```

image

Is there any chance you could update it, I would be interested in taking a look here, but not familiar with the code at all,

thanks!

**Comment by dwnusbaum** [comment]
@timja, sorry, the problem is just that `public class MapParameter ` should be `public static class MapParameter` when you copy it into `DescribableModelTest`. I will update my earlier comment. FWIW, for the questions I asked in https://github.com/jenkinsci/structs-plugin/pull/52#pullrequestreview-338811780, my answers would be:

1. Yes, supporting types such as `MapParameter` is a problem. I think the fix is to do as @daspilker describes [here](https://github.com/jenkinsci/structs-plugin/pull/52#discussion_r315379468).
2. `DescribableModelTest#recursion` should not pass as-is in `master`. `Recursion` should need to implement `Describable` for the test to pass.
    * The implications of making `DescribableModel` more restrictive for `HeteroGeneousObjectType` and `HomogenousObjectType` are unclear. I would run PCT tests with the proposed change to see if any existing types are affected so they can be fixed in advance.

**Comment by jglick** [comment]
I cannot recall now all the reasons why I permitted non-`Describable` types with `@DataBoundConstructor`. I do not think `Recursion` specifically was intended to be testing that loophole, though. The one corner case I can think of is `ParameterValue`, which has non-`Describable` subtypes with `@DataBoundConstructor` on them, critical for example for the Pipeline `build` step. See hacks like `parameterValueClass`. Unless and until the core patch suggested in JENKINS-26093 is in the baseline, you need to retain this special case.

**Comment by zbynek** [comment]
@jglick 

> Because Map is not a legal databinding type.

Is there a reason why `List<String>` is legal and `Map<String, String>` is not?  What would it take to make `Map` legal as well?

As you can see in the [step docs](https://www.jenkins.io/doc/pipeline/steps/jira-steps/) there are multiple steps that take Map as an argument. I was hoping the description of Map and List could be similar in the docs, but the way structs describes Lists and Maps  is different (ArrrayType vs ErrorType).

**Comment by jglick** [comment]
> Is there a reason why `List<String>` is legal and `Map<String, String>` is not?

Neither are legal. Besides atomic-ish types, only `List<SomeKindOfDescribable>` is legal.

**Comment by jglick** [comment]
> there are multiple steps that take Map as an argument

And these plugins are implemented incorrectly.

**Pull request closed without merging**
